confgenerator/resourcedetector/gce_detector.go

```diff
@@ -0,0 +1,125 @@
+package resourcedetector
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Keep synced with the Prometheus repo
+const (
+	metaLabelPrefix       = "__meta_"
+	gceLabel              = metaLabelPrefix + "gce_"
+	gceLabelProject       = gceLabel + "project"
+	gceLabelZone          = gceLabel + "zone"
+	gceLabelNetwork       = gceLabel + "network"
+	gceLabelSubnetwork    = gceLabel + "subnetwork"
+	gceLabelPublicIP      = gceLabel + "public_ip"
+	gceLabelPrivateIP     = gceLabel + "private_ip"
+	gceLabelInstanceID    = gceLabel + "instance_id"
+	gceLabelInstanceName  = gceLabel + "instance_name"
+	gceLabelTags          = gceLabel + "tags"
+	gceLabelMetadata      = gceLabel + "metadata_"
+	gceLabelLabel         = gceLabel + "label_"
+	gceLabelMachineType   = gceLabel + "machine_type"
+	gceLabelInterfaceIPv4 = gceLabel + "interface_ipv4_"
+)
+
+// The data provider interface for GCE environment
+// Implementation of this provider can use either the metadata server on VM,
+// or the cloud API
+type GCEDataProvider interface {
```

Comment:
We should add some comments here around what is available in the detector now and what is not